### PR TITLE
Make it possible to override base package URL location; switch default to GitHub

### LINKED ISSUE #457: version is hard coded so that we can't change the default version in wrap cookbooks
Those codes are evaluated before wrap cookbooks, so we can't change those attributes by overwrite `node['rabbitmq']['version']`.
```ruby:attributes/default.rb
default['rabbitmq']['deb_package'] = "rabbitmq-server_#{node['rabbitmq']['version']}-1_all.deb"
default['rabbitmq']['deb_package_url'] = "https://www.rabbitmq.com/releases/rabbitmq-server/v#{node['rabbitmq']['version']}/"

case node['platform_family']
when 'rhel', 'fedora'
  default['rabbitmq']['rpm_package'] = if node['platform_version'].to_i > 6
                                         "rabbitmq-server-#{node['rabbitmq']['version']}-1.el7.noarch.rpm"
                                       else
                                         "rabbitmq-server-#{node['rabbitmq']['version']}-1.el6.noarch.rpm"
                                       end
when 'suse'
  default['rabbitmq']['rpm_package'] = "rabbitmq-server-#{node['rabbitmq']['version']}-1.suse.noarch.rpm"
end

default['rabbitmq']['rpm_package_url'] = "https://www.rabbitmq.com/releases/rabbitmq-server/v#{node['rabbitmq']['version']}/"
```

Solutions:
---------------

Use `lazy { ... }`
Example:
```
default['rabbitmq']['rpm_package_url'] = lazy { "https://www.rabbitmq.com/releases/rabbitmq-server/v#{node['rabbitmq']['version']}/" }
```

### PR DESCRIPTION
**Pull request by Wing924**
Fixes #457, addresses [RabbitMQ core team request](https://groups.google.com/forum/#!topic/rabbitmq-users/9ASVq55wXnA) to switch to Bintray/GitHub/Package Cloud from rabbitmq.com.

### CODE REVIEW

**Comment by Wing924** [comment]
@michaelklishin All tests are passed except lint test. But it seems not a part of my PR. Please check it.

**Comment by michaelklishin** [comment]
This looks differently from what was originally suggested. @jjasghar how are such attribute initialization issues typically resolved in modern Chef? This approach seems pretty hacky to me.

**Comment by Wing924** [comment]
@michaelklishin I tried `lazy { ... }`, but it don't work with embed string like `"../#{node['rabbitmq'][...']}"`. If you have any better idea, please tell me.

**Comment by yardinicwaller** [comment]
There was already a back-and-forth on this issue back in 3.x. There was a change merged into 3.0.4 which moved URL string interpolation from from the attributes file [into the recipe](https://github.com/rabbitmq/chef-cookbook/commit/65ad07eaed965ab9ed4eb2e8285b86de0779cdfd#diff-25e5d4a4446ae12a0d6f1162b6160375). However, the way it was done broke the ability to override package URL and use a local mirror.

Then later in 3.12.0 that behaviour was reverted and URL string interpolation was moved [back into the attributes file](https://github.com/rabbitmq/chef-cookbook/commit/5aa129516c0d61b664732f50431b64590a1a5ad4#diff-25e5d4a4446ae12a0d6f1162b6160375) which causes the behaviour that @Wing924 is trying to fix.

I believe the best approach is to do URL string interpolation in a recipe, while still allowing that to be overridden by an attribute *if defined*. It would look something like this:

attributes/default.rb:
```
default['rabbitmq']['version'] = '3.5.2'
default['rabbitmq']['deb_package'] = nil # override me if you have a local mirror
```

recipes/default.rb
```
default_deb_package = "https://www.rabbitmq.com/releases/rabbitmq-server/v#{node['rabbitmq']['version']}/"
deb_package = node['rabbitmq']['deb_package'] || default_deb_package
remote_file "#{Chef::Config[:file_cache_path]}/#{deb_package}" do
  source deb_package
```

**Comment by Wing924** [comment]
@michaelklishin 
I updated my PR as @yardinicwaller commented. Can you please check this PR?

File: recipes/default.rb [hunk 1]
```diff
@@ -27,6 +27,29 @@ class Chef::Resource
 
 include_recipe 'erlang'
 
+version = node['rabbitmq']['version']
+
+default_deb_package_name = "rabbitmq-server_#{version}-1_all.deb"
+default_deb_package_url = "https://www.rabbitmq.com/releases/rabbitmq-server/v#{version}/"
```
Comment (michaelklishin) [inline recipes/default.rb]: It's a good opportunity to switch this URL away fro m`rabbitmq.com/releases`, say, to [GitHub releases](https://github.com/rabbitmq/rabbitmq-server/releases). GitHub offers better availability, a CDN and there are not plans to drop it as an artifact distribution mechanism, unlike `rabbitmq.com/releases`.

**Comment by michaelklishin** [comment]
@yardinicwaller thanks for the historical perspective and the suggestion. @Wing924 I'd need to investigate CI failures but let's change default URL to point to GitHub while we are at it. Thank you.

**Comment by michaelklishin** [comment]
CI failures should be addressed in master. @Wing924 please rebase/merge in master.

**Comment by michaelklishin** [comment]
@Wing924 actually, I'm happy to change download URL to point to GitHub or Bintray separately, so let's merge in master to get CI failure fixes and see.

@jjasghar do you have any objections to the approach in this PR?

**Comment by michaelklishin** [comment]
CI in master is green again.

**Comment by michaelklishin** [comment]
@Wing924 there is a [CI failure](https://travis-ci.org/rabbitmq/chef-cookbook/jobs/279820197) that looks genuine:

```
* remote_file[/opt/kitchen/cache/rabbitmq-server_3.6.12-1_all.deb] action create_if_missing[2017-09-26T06:28:25+00:00] WARN: remote_file[/opt/kitchen/cache/rabbitmq-server_3.6.12-1_all.deb] cannot be downloaded from https://github.com/rabbitmq/rabbitmq-server/releases/download/v3.6.12/rabbitmq-server_3.6.12-1_all.deb: 404 "Not Found"
```

The correct URL is https://github.com/rabbitmq/rabbitmq-server/releases/download/rabbitmq_v3_6_12/rabbitmq-server_3.6.12-1_all.deb.

Bintray URLs should require no substitution: https://dl.bintray.com/rabbitmq/rabbitmq-server-deb/rabbitmq-server_3.6.12-1_all.deb.
We have plans to change Bintray repository for Debian packages to just `deb` but we can worry about that when it happens.

**Comment by michaelklishin** [comment]
OK, looks like afcc2863e6a84e797b49055e54e6f06f184b2d18 addresses the above comment.

**Pull request merged**